provisioning/transport/mqtt/src/mqtt.ts

```diff
@@ -4,12 +4,413 @@
 'use strict';
 
 import { EventEmitter } from 'events';
-/*
+import * as uuid from 'uuid';
+import * as machina from 'machina';
+import * as dbg from 'debug';
+const debug = dbg('azure-iot-provisioning-device-mqtt:Mqtt');
+
 import { MqttBase } from 'azure-iot-mqtt-base';
 import { errors, X509 } from 'azure-iot-common';
-import * as machina from 'machina';
-import { DeviceProvisioningTransport } from 'azure-iot-provisioning-device';
-*/
+import { X509ProvisioningTransport } from 'azure-iot-provisioning-device';
+import { ProvisioningDeviceConstants, ProvisioningTransportOptions } from 'azure-iot-provisioning-device';
+import { RegistrationRequest } from 'azure-iot-provisioning-device';
+import { translateError } from 'azure-iot-provisioning-device';
+
+
+/**
+ * Topic to subscribe to for responses
+ */
+const responseTopic: string = '$dps/registrations/res/#';
+
+export class Mqtt extends EventEmitter implements X509ProvisioningTransport {
+  private _mqttBase: MqttBase;
+  private _config: ProvisioningTransportOptions = {};
+  private _fsm: machina.Fsm;
+  private _auth: X509;
+
+  private _operations: {
+    [key: string]: (err?: Error, payload?: any) => void;
+  } = {};
+
+  constructor(mqttBase?: MqttBase) {
+    super();
+    this._mqttBase = mqttBase || new MqttBase(ProvisioningDeviceConstants.apiVersion);
+    this._config.pollingInterval = ProvisioningDeviceConstants.defaultPollingInterval;
+    this._config.timeoutInterval = ProvisioningDeviceConstants.defaultTimeoutInterval;
+
+    const responseHandler = (topic: string, payload: any) => {
+      let payloadString: string = payload.toString('ascii');
+      debug('message received on ' + topic);
+      debug(payloadString);
+
+    /* Codes_SRS_NODE_PROVISIONING_MQTT_18_010: [ When waiting for responses, `registrationRequest` shall watch for messages with a topic named $dps/registrations/res/<status>/?$rid=<rid>.] */
+    /* Codes_SRS_NODE_PROVISIONING_MQTT_18_024: [ When waiting for responses, `queryOperationStatus` shall watch for messages with a topic named $dps/registrations/res/<status>/?$rid=<rid>.] */
+    let match = topic.match(/^\$dps\/registrations\/res\/(.*)\/\?\$rid=(.*)$/);
```

Comment:
>let match = topic.match(/^\$dps\/registrations\/res\/(.*)\/\?\$rid=(.*)$/); [](start = 4, length = 75)

since we're subscribing to only one topic is it really necessary to use a regex? couldn't we just try and split the thing?